### PR TITLE
EM thinning; DINT/EleCa removal

### PR DESCRIPTION
**Pull request by rafaelab**
This pull request introduces thinning for electromagnetic interactions.
It also removes EleCa and DINT to favour the usage of the CRPropa-native electromagnetic modules. These external libraries for electromagnetic studies are no longer supported nor available. 

This also closes issues #199 and #219 .

### CODE REVIEW

File: include/crpropa/module/EMDoublePairProduction.h [hunk 1]
```diff
@@ -29,12 +30,14 @@ class EMDoublePairProduction: public Module {
 	EMDoublePairProduction(
 		PhotonField photonField = CMB, //!< target photon background
 		bool haveElectrons = false,    //!< switch to create the secondary electron pair
+		double thinning = 0,           //!< weighted sampling of secondaries
```
Comment (lukasmerten) [inline include/crpropa/module/EMDoublePairProduction.h:33]: Could we add a comment here that `thinning=0` means basically no thinning?

File: include/crpropa/module/EMInverseComptonScattering.h [hunk 2]
```diff
@@ -34,12 +35,14 @@ class EMInverseComptonScattering: public Module {
 	EMInverseComptonScattering(
 		PhotonField photonField = CMB, //!< target photon background
 		bool havePhotons = false,      //!< switch to create secondary photon
+		double thinning = 0,           //!< weighted sampling of secondaries
```
Comment (lukasmerten) [inline include/crpropa/module/EMInverseComptonScattering.h:38]: Could we add a comment here that `thinning=0` means basically no thinning?

File: include/crpropa/module/EMPairProduction.h [hunk 3]
```diff
@@ -35,12 +36,14 @@ class EMPairProduction: public Module {
 	EMPairProduction(
 		PhotonField photonField = CMB, //!< target photon background
 		bool haveElectrons = false,    //!< switch to create secondary electron pair
+		double thinning = 0,           //!< weighted sampling of secondaries
```
Comment (lukasmerten) [inline include/crpropa/module/EMPairProduction.h:39]: Could we add a comment here that `thinning=0` means basically no thinning?
And in all other modules, too.

File: src/module/EMDoublePairProduction.cpp [hunk 4]
```diff
@@ -8,9 +8,10 @@
 
 namespace crpropa {
 
-EMDoublePairProduction::EMDoublePairProduction(PhotonField photonField, bool haveElectrons, double limit) {
+EMDoublePairProduction::EMDoublePairProduction(PhotonField photonField, bool haveElectrons, double thinning, double limit) {
```
Comment (lukasmerten) [inline src/module/EMDoublePairProduction.cpp:11]: Do old scripts, using only photonField, haveElectrons, and limit as arguments still run?

File: include/crpropa/module/EMPairProduction.h (hunk 3 above)
Comment (rafaelab) [reply include/crpropa/module/EMPairProduction.h:39]: Good idea! I've also added a better description of the thinning to the comments in the header files.

File: src/module/EMDoublePairProduction.cpp (hunk 4 above)
Comment (rafaelab) [reply src/module/EMDoublePairProduction.cpp:11]: The old scripts run as usual for the photonField and haveElectrons arguments. But I had to make a decision concerning the limit argument. Since few people set the limit anyway, I added the thinning argument BEFORE the limit to make it more accessible. This would break scripts that explicitly use this argument.

**Comment by rafaelab** [comment]
The travis is not building because make test is not working.
I have to make a few adjustments. I believe this is because there is an inconsistency in the limiting of step sizes.

File: src/module/EMDoublePairProduction.cpp [hunk 5]
```diff
@@ -69,8 +75,19 @@ void EMDoublePairProduction::performInteraction(Candidate *candidate) const {
 	Random &random = Random::instance();
 	Vector3d pos = random.randomInterpolatedPosition(candidate->previous.getPosition(), candidate->current.getPosition());
 
-	candidate->addSecondary( 11, Ee, pos);
-	candidate->addSecondary(-11, Ee, pos);
+	double f = Ee / E;
+	double w0 = candidate->getWeight();
+
+	if (haveElectrons) {
```
Comment (lukasmerten) [inline src/module/EMDoublePairProduction.cpp:81]: Does the difference in the two if conditions -- (1-f)^\alpha compared to f^alpha -- introduce a difference between particles and anti-particles or am I missing something?

When the energy fraction f is small then electrons are produced more often than positrons, aren't they?

Comment (rafaelab) [reply src/module/EMDoublePairProduction.cpp:81]: It does. It increases efficiency, but since we are not tracking all the particles charge is not conserved.

File: src/module/EMDoublePairProduction.cpp (hunk 4 above)
Comment (lukasmerten) [reply src/module/EMDoublePairProduction.cpp:11]: I agree that most people do not set these arguments but would it be possible to add a meaningfull warning, when these arguments are set? 

If not the user will probably figure out fast that a new arguments was introduced and is able to adopt his code based on the default error message.

Comment (rafaelab) [reply src/module/EMDoublePairProduction.cpp:11]: @lukasmerten : do you think this is a problem?
Do you use the limit argument?

File: src/module/EMPairProduction.cpp [hunk 6]
```diff
@@ -190,11 +197,20 @@ void EMPairProduction::performInteraction(Candidate *candidate) const {
...
 	double Ep = E - Ee;
+	double f = Ep / E;
 
 	// sample random position along current step
 	Vector3d pos = random.randomInterpolatedPosition(candidate->previous.getPosition(), candidate->current.getPosition());
-	candidate->addSecondary(-11, Ee / (1 + z), pos);
-	candidate->addSecondary(11, Ep / (1 + z), pos);
+	double w0 = candidate->getWeight();
+	// apply sampling
+	if (random.rand() < pow(f, thinning)) {
```
Comment (lukasmerten) [inline src/module/EMPairProduction.cpp:206]: Here the probabilities are now the opposite for electrons and positrons compared to EMDoublePairProduction, right?

Is this by chance or is there a reason for it; making up for some of the charge imbalence?

Comment (rafaelab) [reply src/module/EMPairProduction.cpp:206]: The convention doesn't really matter. The total energy is fixed. If the electron takes a fraction f of the primary energy (E), then it's energy is Ee, such that Ee+Ep=E.
If we choose a random number r and accept electrons if r<f^thinning, positrons are accepted if r<(1-f)^thinning. The only important requirement is to keep this consistency that for one particle we have f, for the other 1-f.

File: src/module/EMDoublePairProduction.cpp (hunk 4 above)
Comment (lukasmerten) [reply src/module/EMDoublePairProduction.cpp:11]: No I don't think this is a problem and I am also not setting this values explicitly. So it is probably fine as it is.

File: src/module/EMPairProduction.cpp (hunk 6 above)
Comment (rafaelab) [reply src/module/EMPairProduction.cpp:206]: Electrons or positrons are accepted based on random numbers, not on charge. The particle with the highest energy fraction tends to be added to the list. For thinning=1, if the electron takes 90% of the energy, it is likely that it will be added to the stack. In this case, the positron takes 10% of the energy, so it's less likely that it will be accepted.
If, on the other hand, the positron takes 90% of the energy and the electron 10%, it is the other way around.

File: src/module/EMDoublePairProduction.cpp (hunk 5 above)
Comment (lukasmerten) [reply src/module/EMDoublePairProduction.cpp:81]: Would it be possible to just dice the sign of the particle ID?
In this way the efficiency should stay the same but on average for constant energy fractions f the number of positrons and electrons should converge to the same value.

Comment (rafaelab) [reply src/module/EMDoublePairProduction.cpp:81]: Take a look at my comment below. There is no bias towards either charge, since the ones that are kept are chosen randomly based solely on the fraction of the total energy they take. The more energetic ones are kept, and their weights increased, whereas the low-energy ones will rarely be added to the array of secondaries.
If electrons and positrons have equal probability of being produced, then two random numbers are drawn and, on average, the simulation will be charge-neutral.
In other words, the higher the thinning, the more the most energetic particles are valued, since there are only a few of them compared to the many that are produced at lower energies.

Comment (lukasmerten) [reply src/module/EMDoublePairProduction.cpp:81]: I agree with you that is is true for the the other module but here I am not yet convinced.

Let's assume we have a maximal thinning `thinning=1` and the primary gamma-ray has an energy `E=4*m_e c^2 ` -- which might not be realistic due to production thresholds but makes my argument more clear.

In this case, the energy for positron and electron is `Ee=(E - 2*m_e c^2)/2=m_e c^2 ` and the energy fraction is `f=1/4`. 

Then we would have a 75%=(1-f) chance of producing an electron and a 25% chance for a positron. When I read the code correctly, these chances are constant for a given input energy E. 

This is different from the other module, where Ee!=Ep and Ee will be different in each call of the performInteraction function.

One last thought:
In reality it's possible that this treatment is not problematic at all.
Assuming that `E>>m_e c^2` --> f=1/2 which will lead to a equal production rate for positrons and electrons.

Comment (rafaelab) [reply src/module/EMDoublePairProduction.cpp:81]: This is not exactly true.
This implementation follows Lee 1996 mentioned in the comments. He says:
"Therefore, I use the assumption where one pair of the two carries all the initial energy and two particles in the pair share the energy equally."
Therefore, f=1/2 (not 1/4) for one electron and positron, and 0 for the others.

This is weird, I agree. DPP in this case is just the usual PP (it only produces one pair, effectively), but the cross sections are different, hence it's importance. There might be better implementations.

I could have just set f=0.5 from the beginning. I kept it this way because in the future we might want to improve this implementation and the it will require less work.

And this treatment is okay exactly for the reason you said: f=1/2 at high energies, so regardless of the thinning we get + and - charges balanced.

Comment (lukasmerten) [reply src/module/EMDoublePairProduction.cpp:81]: @rafaelab Thanks for you detailed explanations. I think I just misunderstood the implementation regardless of the thinning or not. I should have taken a look into the original paper (Lee 1996). I'll try to come up with a short comment that we might add to the code, to make the procedure more clear. But this should not block your PR.

**Comment by lukasmerten** [comment]
In addition to all the smaller comments, I am wondering if we want to drop DINT and EleCa with this PR or wait until the 3.2 release. Here, I am actually quite open minded and would like to read your thoughts.

File: src/module/EMDoublePairProduction.cpp (hunk 5 above)
Comment (rafaelab) [reply src/module/EMDoublePairProduction.cpp:81]: But now that you mention, although they are the same in this case, one line should be Ee, and the other Ep, for consistency. I will change that.

**Comment by rafaelab** [comment]
> In addition to all the smaller comments, I am wondering if we want to drop DINT and EleCa with this PR or wait until the 3.2 release. Here, I am actually quite open minded and would like to read your thoughts.

That's a good point. On one hand, we are no longer offering support for DINT, so it makes sense to provide an efficient alternative. On the other hand, it makes sense to do all this when 3.2 is ready. Therefore, it might be a good idea to wait to merge this pull request when releasing 3.2.

**Comment by rafaelab** [comment]
This pull request will be closed tomorrow. It will be split in two: the thinning part, and the removal of DINT/EleCa.
